ambry-network/src/main/java/com.github.ambry.network/BlockingChannelConnectionPool.java

```diff
@@ -223,6 +223,15 @@ public void destroyBlockingChannel(BlockingChannel blockingChannel) {
       synchronized (lock) {
         // decrement the number of connections to the host and port. we were not able to maintain the count
         numberOfConnections.decrementAndGet();
+        // at this point we are good to clean up the available connections since re-creation failed
+        do {
```

Comment:
Deleted an earlier comment about the data structure's thread safety - it is not concurrent, but it is thread-safe. I think the `do-poll()` should be sufficient here.

@nsivabalan regardless, I suggest getting clarity on what exactly the purposes of `rwLock` and the `lock` object are.